src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs

```diff
@@ -335,7 +335,13 @@ private void ReportMismatchInAttributesDiagnostic (SymbolAnalysisContext symbolA
 				message));
 		}
 
-		private bool HasMismatchingAttributes (ISymbol member1, ISymbol member2) => member1.HasAttribute (RequiresAttributeName) ^ member2.HasAttribute (RequiresAttributeName);
+		// TODO: Should consider when a symbol is a method of a class that has the attribute
+		private bool HasMismatchingAttributes (ISymbol member1, ISymbol member2)
+		{
+			bool member1HasAttribute = IsMemberInRequiresScope (member1);
```

Comment:
It looks like this should be `TargetHasRequiresAttribute`, not `IsMemberInRequiresScope`, but it looks like `TargetHasRequiresAttribute` may need more work to handle associated symbols (properties)

Reply:
Why do you think it should be `TargetHasRequires`? In the linker we check the equivalent of `IsMemberInRequiresScope`.
```csharp
class Base {
  [RUC]
  virtual void M() {} // TargetHasRequires = true, IsInRequiresScope = true
}

[RUC]
class Derived : Base {
  override void M() {} // TargetHasRequires = false, IsInRequiresScope = false
}
```

I agree that `IsMemberInRequiresScope` needs to be updated to handle properties.

However this does raise an interesting question - what happens in the following case?
```csharp
[RUC]
class Base {
  virtual void M() {} // TargetHasRequires = false, IsInRequiresScope = true
}

class Derived : Base {
  [RUC]
  override void M() {} // TargetHasRequires = true, IsInRequiresScope = true
}
```

I think this shouldn't warn about a mismatch, similar to above. The justification is that we don't need to warn for calls to `Base.M` even though it might require unreferenced code, because constructing the `Base` type produces a warning already. But now you can do `new Derived().M()`. In IL this becomes a call to `Base.M`, which doesn't warn, but at runtime it calls the derived method. Does the analyzer produce any warnings for this? In the linker it's ok since the derived ctor will produce a warning (IIRC) about the call to the base ctor (which has TargetHasRequires = true). But in the analyzer there's no location for the compiler-generated derived ctor.

Do we need to ensure that compiler-generated derived ctors warn if the base class has RUC or are we already handling this somehow?

Reply:
I didn't realize `IsMemberInRequiresScope` is what's used in the linker. My mistake. In that case, we should use that instead.

Reply:
Per discussion I think the concern about compiler-generated derived ctors will be addressed by https://github.com/dotnet/linker/issues/2416.